Feature / add DK structure slash enable easier adding of new markets/locales

styrelius opened the pull request:
### What?
This PR suggests new ways of adding and accessing values connected to a specific locale, like `hrefLang` and market names for example:
1) In `locales.ts` there is a big object containing one object for every possible locale with all values that differ between markets/locales. The one object with the current locale is then saved as context (through `constate`) and can be accessed throughout the repo. The other, not current, locales are used in order to get the right values for the `LanguagePicker` and `MarketPicker`.
2) There is also one addition to Storyblok `GlobalStory`: `four_oh_four_title`. This is used in the `FourOhFourBlock`.

In addition to this a few UX tweaks are made in `LanguagePicker` and `MarketPicker` - like disabling clicks on the current active language link in `LanguagePicker` and setting value of the `MarketSelect` in `MarketPicker` to the current market since the value of the dropdown isn't used to redirect and there was this strange behaviour where it looked like you'd change market but there was no redirect to that market specific version of the site. 

### Why?
The previous solution to locale specific data was changing several hard coded strings in conditional statements in a bunch of different files. This doesn't really scale so well and the changes suggested here will make it a lot easier when going to a new market since there will only be one place where stuff needs to be added in conjunction with creating a new dimension in Storyblok and this one place is the `locales.ts` file.

### GIFs
![2020-09-23-languagepicker](https://user-images.githubusercontent.com/42962286/94014468-db60a380-fdab-11ea-9d17-a3cd4248d6fa.gif)

![2020-09-23-marketpicker-no](https://user-images.githubusercontent.com/42962286/94014788-4dd18380-fdac-11ea-9375-1b7f60b2b6b9.gif)

![2020-09-23-marketpicker-da](https://user-images.githubusercontent.com/42962286/94014599-0fd45f80-fdac-11ea-9843-1cc2274dde74.gif)

Commit:
Add danish start page redirect

styrelius committed:
Replace all occurrences of 'lang' with 'locale'. Start fetching locale from Storyblok global story. Add tests for getLocaleFromPath.

styrelius committed:
Fetch FourOhFour copy from GlobalStory instead of having hard coded strings in code base.

styrelius committed:
Delete /HeaderBlockBrandPivot/MarketPicker since it's not used anywhere.

styrelius committed:
Add html_lang to Storyblok Global story Locale property. Use this to get the right value for html lang tag in /server/page.tsx. Also add destructuring of story object in <GlobalStoryContainer> in FourOhFourBlock.

styrelius committed:
Start remake of how we work with locale specific data, in order for us to only make additions in one single place when we go to a new market:
Create an exported "locales" object containing one object for every locale with all locale specific data needed (like label, hrefLang etc.) and an exported getLocaleData function in CurrentLocale.tsx, then use these to add a currentLocale object to context, instead of just a "locale" string.
Use the currentLocale context and getLocaleData to get correct hrefLang tags etc. in meta.tsx and replace all use of old context locale with this new shiny one. Also try to fix stuff that broke along the way :)

styrelius committed:
Add langLabel property to the locale objects in CurrentLocale.tsx. Also add a getAssociatedLocales function. Use these changes to make the LanguagePicker component generic and not relying on hard coded strings to display the right values.

styrelius committed:
Remake MarketPicker.tsx so that it gets its values based on CurrentLocale context and the locales in CurrentLocale.tsx. Add some functions (as well as remove some) in CurrentLocale.tsx in order to achieve this.

styrelius committed:
Make minor modifications to getMarketsInLocalLang and getMarketsInEnglish functions in CurrentLocale.tsx.

styrelius committed:
Change name of CurrentLocal.tsx to locales.ts, and also change page test file extension from .ts to .tsx.

styrelius committed:
Make minor fixes/clean up a little in meta.tsx and MarketPicker.tsx.

styrelius committed:
Add constate Providers to FooterBlock.stories and HeaderBlock.stories with fallbackLocale passed in as currentLocale.

styrelius committed:
Merge branch 'master' into feature/dk-structure

styrelius committed:
Remove the html_lang property from Storyblok global story since it makes more sense to only have stuff there that you actually see in the ui. Add an htmlLang property to all the locales objects in locales.ts and change so that html lang in template (page.jsx) comes from context.currentLocale instead of from Storyblok globalStory.

styrelius committed:
Upgrade @types/react-select

styrelius committed:
Fix duplicated _uid in footerMenuItems in storybook.ts.

styrelius committed:
Pass FooterBlock _uid to MarketPicker as blokId prop, in order to pass it to the select component as instanceId :)
Remove getLocalesWithoutCurrent function from MarketPicker component since it seems like the defaultValue also needs to be among the options. Add a function to sort the markets alphabetically in the same component.

styrelius committed:
Add value to MarketSelect in MarketPicker.tsx so that it always displays the marketName of currentLocale. Fix market options values to make the routing reasonable.

styrelius committed:
Fix graphql query that I accidently screwed up in usePerils.ts.

styrelius committed:
Pass PerilsBlock _uid to SelectInsurance as instanceId.

palmenhq commented on src/components/containers/ContextContainer.tsx:
```diff
@@ -1,8 +1,9 @@
 
-interface Context {
-  lang: string
+type Context = {
+  currentLocale: LocaleData
 }
```
Maybe we should use `interface` still?

styrelius replied:
I'm going to leave the `type` alias since we agreed that we might transition to using `type` over `interface` for component props and state, and that it's fine that both are used in the codebase for now. 😇

palmenhq reviewed (dismissed):
Really REALLY really well done!
Just a couple of optional naming suggestions

styrelius commented on src/blocks/FooterBlock/MarketPicker/MarketPicker.tsx:
```diff
@@ -13,31 +19,47 @@ const MarketSelect = styled(Select)`
 }) => {
-  const marketLabel = getMarketLabel(currentLanguage)
-  const activeMarket = getCurrentMarket(currentLanguage)
+  const currentMarket = currentLocale.marketName
```
Yup, that's better, will change that! 👍

styrelius committed:
Change name of currentMarket to currentMarketaName in MarketPicker.tsx.

styrelius committed:
Replace hard coded if statements with using locales object
in order to make the redirects based on IP address in startPageRedirect in redirects.ts.

gustaveen commented on src/server/middlewares/redirects.ts:
```diff
@@ -26,20 +27,16 @@ export const forceHost = ({
-  if (country === 'NO') {
-    ctx.redirect(`/no${queryStringMaybe}`)
-    return
+  if (Object.keys(locales).includes(countryLabel)) {
+    ctx.redirect(`/${countryLabel}${queryStringMaybe}`)
+  } else {
+    ctx.redirect(`/${fallbackCountryLabel}${queryStringMaybe}`)
   }
```
Maybe just have a early return? :) 
```suggestion
  if (Object.keys(locales).includes(countryLabel)) {
    ctx.redirect(`/${countryLabel}${queryStringMaybe}`)
    return
  }
  
  ctx.redirect(`/${fallbackCountryLabel}${queryStringMaybe}`)

```

styrelius replied:
Sure, I don't have very strong opinions on using the `else` 🙃

gustaveen commented:
Amazing job with this Ellen! Good to release IMO! 🥳

styrelius committed:
Remove else statement and add early return in startPageRedirect func in redirects.ts since Gustav asked me to :)

styrelius committed:
Merge branch 'master' into feature/dk-structure

gustaveen reviewed (approved):
Let's 🚢   it!
